gcovr/__main__.py

```diff
@@ -472,13 +481,16 @@ def main(args=None):
 
     # Get coverage data
     covdata = {}
+    pool = Workers(options.gcov_parallel)
+    if options.verbose:
+        print("Pool started with %d workers" % pool.size())
     for file_ in datafiles:
         if options.gcov_files:
-            process_existing_gcov_file(file_, covdata, options)
+            pool.add(process_existing_gcov_file, file_, covdata, options)
         else:
-            process_datafile(file_, covdata, options)
+            pool.add(process_datafile, file_, covdata, options)
```

Comment:
The `covdata` dict will be updated from multiple threads, but is not synchronized. Should this access be guarded through a mutex section? Or should the workers write update commands to an output queue that is then executed on the main thread?

There may also be concurrent accesses to one covdata entry if multiple threads provide coverage for the same file, in particular for header files that are included in multiple parts of the project.

Reply:
Good remark, access to `covdata` shall be synchronized. One other possibility not mentioned by @latk is to have one `covdata` dict per thread and do a threaded merge at the end (nb thread/2 at each merge step).

Reply:
I've created a LockedDictionary object that updates with a lock. As this is only called in one place after the recent refactor to gcov.py this seemed to be the easiest option.

However... keeping a context per thread and merging is a nice pattern and avoids further locks. It also may cure help the problem in the next comment...

Reply:
Just forget my comment regarding threaded merge. This won't  do any good running only python code. I should have known better as I always try to release the GIL in long running tasks in C extensions to allow multithreading... Python only runs python code in one thread at a time.

Reply:
@mayeut Just to clarify, you're saying “multi-level merging is pointless in Python because merging can't happen in parallel”, but not “collecting separately and merging at the end is unnecessary”, right? So the current design (with a per-thread `covdata` dictionary) is correct as per your experience?

Reply:
Right.

Reply:
@latk, I really need to think "python". Thinking again about your question. I did not answer correctly. So, the only thing that can benefit from threading are I/O here (subprocess being part of that). Updating the coverage dictionary is pure python with no I/O, 2 threads can't really do that in parallel (compute time wise) so, we're probably better off protecting access to the dictionary rather than merging at the end since there can't really be contention here. Merging can be easier (thread safety) but not faster.